src/compiler/checker.ts

```diff
@@ -32200,19 +32200,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getSpreadArgumentType(args: readonly Expression[], index: number, argCount: number, restType: Type, context: InferenceContext | undefined, checkMode: CheckMode) {
+        const inConstContext = isConstTypeVariable(restType);
+
         if (index >= argCount - 1) {
             const arg = args[argCount - 1];
             if (isSpreadArgument(arg)) {
                 // We are inferring from a spread expression in the last argument position, i.e. both the parameter
                 // and the argument are ...x forms.
-                return getMutableArrayOrTupleType(arg.kind === SyntaxKind.SyntheticExpression ? (arg as SyntheticExpression).type :
-                    checkExpressionWithContextualType((arg as SpreadElement).expression, restType, context, checkMode));
+                const spreadType = arg.kind === SyntaxKind.SyntheticExpression ? (arg as SyntheticExpression).type :
+                    checkExpressionWithContextualType((arg as SpreadElement).expression, restType, context, checkMode);
+
+                if (isArrayLikeType(spreadType)) {
+                    return getMutableArrayOrTupleType(spreadType);

```

Comment:
likely this shouldn't always produce mutable types **but** there are other bugs here, like this one:
```ts
declare function foo<const T extends readonly unknown[]>(...args: T): T;
const p = foo(...[1, 'bar']); // A spread argument must either have a tuple type or be passed to a rest parameter.(2556)
const p2 = foo(true, ...[1, 'bar']); // OK
```

I would prefer to fix this in a followup PR as the fix for this largely lies elsewhere - then I could revisit this piece of code to produce mutable/readonly array/tuple types based on `inConstContext`.

Sideways q - should `T` in the broken case be inferred as `readonly (number | string)[]` or should it be `readonly [1, 'bar']`? In other words - should the spread array expression inherit the const context?

Reply:
> Sideways q - should `T` in the broken case be inferred as `readonly (number | string)[]` or should it be `readonly [1, 'bar']`? In other words - should the spread array expression inherit the const context?

I _think_ the latter?

Reply:
The justification in my mind being that it should match what you get if you write `foo(1, 'bar')`.

Reply:
I also think that the latter is better but currently it's complicated to implement it. IIRC the easiest way to achieve this would be through `getEffectiveCallArguments` but it creates a cycle for spreads. `getContextualTypeForArgument`/`getContextualTypeForArgumentAtIndex` call back into `getEffectiveCallArguments`. I think that I've even seen a comment somewhere about spread arguments never needing the contextual type - so this would be the first time when that would actually have to become supported.

Reply:
Yeah, maybe. I can't say that I explored this in full yet. I might have also missed some additional things from my initial investigation, it's a little bit stale in my memory ;p 

I could try to give this a shot in a followup PR unless you strongly think that it should go in together.

Reply:
Unless I'm mistaken, your code sample behaves the same between `main` and this PR, so, I feel like it's fine to leave as-is and then report a bug for further refinement.

Reply:
(But do tell me if I'm right or wrong here)

Reply:
> Unless I'm mistaken, your code sample behaves the same between main and this PR,

Ye, the behavior is the same.

> I feel like it's fine to leave as-is and then report a bug for further refinement.

It seems that I already started to work on further improvements related to this [here](https://github.com/microsoft/TypeScript/pull/52845) (who could have guessed that? 😱 ) and I even mentioned some more details around that issue with `getEffectiveCallArguments` in the comment [here](https://github.com/microsoft/TypeScript/pull/52845#discussion_r1111120187)

Reply:
Ah, I see your memory is as stale as mine is on a regular basis